### PR TITLE
Fix Issue #4502 - Confusing exception when trying to map types with …

### PR DESCRIPTION
**Pull request by Biotronic**
…`in` parameters in ctor

See issue #4502.

### CODE REVIEW

**Comment by Biotronic** [comment]
@dotnet-policy-service agree

**Comment by Biotronic** [comment]
@dotnet-policy-service agree

**Comment by lbargaoanu** [comment]
Revart _all_ the whitespace changes. Assert that the test mapping actually works.

File: src/AutoMapper/Internal/TypePair.cs [hunk 1]
```diff
@@ -11,15 +11,17 @@ public readonly record struct TypePair(Type SourceType, Type DestinationType)
 {
     public bool IsConstructedGenericType => SourceType.IsConstructedGenericType || DestinationType.IsConstructedGenericType;
     public bool ContainsGenericParameters => SourceType.ContainsGenericParameters || DestinationType.ContainsGenericParameters;
+    public Type DestinationType { get; init; } = DestinationType.IsByRef ? DestinationType.GetElementType() : DestinationType;
```
Comment (lbargaoanu) [inline src/AutoMapper/Internal/TypePair.cs:14]: This is not an acceptable overhead. `TypePair` is used everywhere. This check should only appeaer only in constructor mapping code.

File: src/AutoMapper/ConstructorMap.cs [hunk 2]
```diff
@@ -80,20 +80,20 @@ public ConstructorParameterMap(TypeMap typeMap, ParameterInfo parameter, MemberI
         }
     }
     public ParameterInfo Parameter { get; }
-    public override Type DestinationType => Parameter.ParameterType;
+    public override Type DestinationType => Parameter.ParameterType.IsByRef ? Parameter.ParameterType.GetElementType() : Parameter.ParameterType;
```
Comment (lbargaoanu) [inline src/AutoMapper/ConstructorMap.cs:83]: Let's do this once in the constructor.

File: src/UnitTests/Bug/ByrefConstructorParameter.cs [hunk 3]
```diff
@@ -0,0 +1,36 @@
+﻿namespace AutoMapper.UnitTests.Bug
+{
+    namespace ByrefConstructorParameter
```
Comment (lbargaoanu) [inline src/UnitTests/Bug/ByrefConstructorParameter.cs:3]: Let's remove the extra namespace. And use file scoped namespaces.

File: src/AutoMapper/Internal/TypePair.cs [hunk 4]
```diff
@@ -7,10 +7,19 @@ public MapRequest(TypePair types) : this(types, types, MemberMap.Instance) { }
     public override int GetHashCode() => HashCode.Combine(RequestedTypes, RuntimeTypes);
 }
 [DebuggerDisplay("{SourceType.Name}, {DestinationType.Name}")]
-public readonly record struct TypePair(Type SourceType, Type DestinationType)
+public readonly record struct TypePair
 {
+    public TypePair(Type sourceType, Type destinationType)
+    {
+        SourceType = sourceType;
+        DestinationType = destinationType.IsByRef ? destinationType.GetElementType() : destinationType;
```
Comment (lbargaoanu) [inline src/AutoMapper/Internal/TypePair.cs:15]: No, what I meant, no change is ok here, in this class. Only constructor mapping code can change.

File: src/AutoMapper/ConstructorMap.cs [hunk 5]
```diff
@@ -70,6 +70,7 @@ public class ConstructorParameterMap : MemberMap
     public ConstructorParameterMap(TypeMap typeMap, ParameterInfo parameter, MemberInfo[] sourceMembers) : base(typeMap)
     {
         Parameter = parameter;
+        DestinationType = parameter.ParameterType.IsByRef ? parameter.ParameterType.GetElementType() : parameter.ParameterType;
```
Comment (lbargaoanu) [inline src/AutoMapper/ConstructorMap.cs:73]: Extract `parameter.ParameterType` to a local.

**Comment by lbargaoanu** [comment]
The tests seem to fail. In order to run them on CI you nees to rebase on  seal_mappers.

**Comment by Biotronic** [comment]
Works beautifully. :)

**Pull request merged**